packages/gatsby/src/schema/infer/add-inferred-fields.js

```diff
@@ -293,18 +231,26 @@ const getFieldConfigFromFieldNameConvention = ({
   // (ii) hinders reusing types.
   if (linkedTypes.length > 1) {
     const typeName = linkedTypes.sort().join(``) + `Union`
-    type = schemaComposer.getOrCreateUTC(typeName, utc => {
-      const types = linkedTypes.map(typeName =>
-        schemaComposer.getOrCreateOTC(typeName)
-      )
-      utc.setTypes(types)
-      utc.setResolveType(node => node.internal.type)
+    type = UnionTypeComposer.createTemp({
+      name: typeName,
+      types: () => linkedTypes.map(typeName => schemaComposer.getOTC(typeName)),
     })
+    type.setResolveType(node => node.internal.type)
   } else {
     type = linkedTypes[0]
   }
 
-  return { type, resolve: link({ by: foreignKey || `id` }) }
+  return {
+    type,
+    extensions: {
```

Comment:
We should also use the extension in `getFieldConfigFromMapping`

Reply:
thanks